diff -rupNb gcc-5.3.0-mmi/config.sub gcc-5.3.0-full/config.sub
--- gcc-5.3.0-mmi/config.sub	2015-01-02 17:30:21.000000000 +0800
+++ gcc-5.3.0-full/config.sub	2016-04-03 17:07:06.847986617 +0800
@@ -765,6 +765,10 @@ case $basic_machine in
 		basic_machine=i386-unknown
 		os=-vsta
 		;;
+	iop)
+		basic_machine=mipsel-scei
+		os=-irx
+		;;
 	iris | iris4d)
 		basic_machine=mips-sgi
 		case $os in
@@ -1393,7 +1397,8 @@ case $os in
 	      | -os2* | -vos* | -palmos* | -uclinux* | -nucleus* \
 	      | -morphos* | -superux* | -rtmk* | -rtmk-nova* | -windiss* \
 	      | -powermax* | -dnix* | -nx6 | -nx7 | -sei* | -dragonfly* \
-	      | -skyos* | -haiku* | -rdos* | -toppers* | -drops* | -es* | -tirtos*)
+	      | -skyos* | -haiku* | -rdos* | -toppers* | -drops* | -es* | -tirtos* \
+	      | -irx*)
 	# Remember, each alternative MUST END IN *, to match a version number.
 		;;
 	-qnx*)
diff -rupNb gcc-5.3.0-mmi/configure gcc-5.3.0-full/configure
--- gcc-5.3.0-mmi/configure	2015-05-04 01:29:57.000000000 +0800
+++ gcc-5.3.0-full/configure	2016-04-03 17:07:06.851986617 +0800
@@ -3402,6 +3402,8 @@ case "${target}" in
     ;;
   mips*-*-linux*)
     ;;
+  mipsel-*-irx*)
+    ;;
   powerpc*-*-linux*)
     ;;
   powerpc-*-darwin*)
diff -rupNb gcc-5.3.0-mmi/configure.ac gcc-5.3.0-full/configure.ac
--- gcc-5.3.0-mmi/configure.ac	2015-05-04 01:29:57.000000000 +0800
+++ gcc-5.3.0-full/configure.ac	2016-04-03 17:07:06.855986616 +0800
@@ -749,6 +749,8 @@ case "${target}" in
     ;;
   mips*-*-linux*)
     ;;
+  mipsel-*-irx*)
+    ;;
   powerpc*-*-linux*)
     ;;
   powerpc-*-darwin*)
diff -rupNb gcc-5.3.0-mmi/gcc/config.gcc gcc-5.3.0-full/gcc/config.gcc
--- gcc-5.3.0-mmi/gcc/config.gcc	2015-09-10 22:17:53.000000000 +0800
+++ gcc-5.3.0-full/gcc/config.gcc	2016-04-03 17:07:06.855986616 +0800
@@ -2161,6 +2161,11 @@ mips64r5900-*-elf* | mips64r5900el-*-elf
 	tmake_file="mips/t-elf"
 	tm_defines="${tm_defines} MIPS_ISA_DEFAULT=3 MIPS_ABI_DEFAULT=ABI_N32"
 	;;
+mipsel-*-irx*)
+	tm_file="dbxelf.h elfos.h newlib-stdint.h ${tm_file} mips/elf.h"
+	tmake_file="mips/t-elf mips/dbx.h"
+	target_cpu_default=MASK_SOFT_FLOAT_ABI
+	;;
 mips64-*-elf* | mips64el-*-elf*)
 	tm_file="elfos.h newlib-stdint.h ${tm_file} mips/elf.h"
 	tmake_file="mips/t-elf"
diff -rupNb gcc-5.3.0-mmi/libgcc/config.host gcc-5.3.0-full/libgcc/config.host
--- gcc-5.3.0-mmi/libgcc/config.host	2016-03-27 16:02:44.000000000 +0800
+++ gcc-5.3.0-full/libgcc/config.host	2016-04-03 17:07:06.859986617 +0800
@@ -889,6 +889,9 @@ mips-*-elf* | mipsel-*-elf*)
 	fi
 	extra_parts="$extra_parts crti.o crtn.o"
 	;;
+mipsel-*-irx)
+	tmake_file="$tmake_file mips/t-elf"
+	;;
 mipsr5900-*-elf* | mipsr5900el-*-elf*)
 	tmake_file="$tmake_file mips/t-elf mips/t-crtstuff"
 	extra_parts="$extra_parts crti.o crtn.o"
